src/Http/ServerRequest.php

```diff
@@ -515,11 +515,39 @@ public function contentType()
     }
 
     /**
+     * Update the request with a new session instance.
+     *
+     * Returns an updated request object. This method returns
+     * a *new* request object and does not mutate the request in-place.
+     *
+     * @param \Cake\Network\Session $session the session object to use
+     * @return static
+     */
+    public function withSession(Session $session)
+    {
+        $new = clone $this;
+        $new->session = $session;
```

Comment:
Sessions are funny because they are effectively shared global state. I don't think we can or should try to make promises around sessions and request immutability.